fil-proofs-tooling/scripts/retry.sh

```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+
+# This software is freely available under the Apache 2.0 license.
+# Full license available here: https://www.apache.org/licenses/LICENSE-2.0.txt
+
+# Inspired by https://gist.github.com/reacocard/28611bfaa2395072119464521d48729a
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# Retry a command on a particular exit code, up to a max number of attempts,
+# with exponential backoff.
+# Invocation:
+#   err_retry exit_code attempts sleep_multiplier <command>
+# exit_code: The exit code to retry on.
+# attempts: The number of attempts to make.
+# sleep_millis: Multiplier for sleep between attempts. Examples:
+#     If multiplier is 1000, sleep intervals are 1, 2, 4, 8, 16, etc. seconds.
+#     If multiplier is 5000, sleep intervals are 5, 10, 20, 40, 80, etc. seconds.
+
+exit_code=$1
+attempts=$2
+sleep_millis=$3
+shift 3
+
+for attempt in `seq 1 $attempts`; do
+    # This weird construction lets us capture return codes under -o errexit
+    "$@" && rc=$? || rc=$?
+
+    if [[ ! $rc -eq $exit_code ]]; then
+        exit $rc
+    fi
+
+    if [[ $attempt -eq $attempts ]]; then
+        exit $rc
+    fi
+
+    sleep_ms="$(($attempt * $attempt * $sleep_millis))"
```

Comment:
This might be a fine backoff strategy, but it's not exponential as documented above, as far as I can tell. This seems to back off  by the square of the attempt. I would either change the implementation or the documentation. For example, the more accurate documentation would describe the first set of sleep intervals as `1, 4, 9, 16, 25`.

I also have to say that the precision/accuracy mismatch between the documentation and implementation do not inspire confidence in the source of this code snippet. It's likely not a major issue for this glue code, but as a matter of principle I'm concerned about introducing fuzziness into the `rust-fil-proofs` code base — so let's at least make it consistent.

Reply:
If you actually want the sleep time to double on each iteration, you can repeatedly assign to a scale variable, doubling it each time, and multiply that by `$sleep_millis`.

Reply:
> This might be a fine backoff strategy, but it's not exponential as documented above, as far as I can tell.

That's fair. The source of the discrepancy between the documentation and implementation (specifically the `If multiplier is 1000, sleep intervals are` section) is me. I changed the original implementation but did not update the copied documentation.

Reply:
If you say so. When I look back at the original it *appears* to have the same problem. I mention this only because maybe we're looking at something different if you think you introduced the issue.